### ISSUE TITLE
Updating Two Glyphs using Periodic Callback

### ISSUE THREAD

**Issue by seanlaw** [bokeh/bokeh#4911]
- **Description of expected behavior and the observed behavior**

Essentially, I'm trying to draw a ball that is zig-zagging downward down ten steps. To show its path, I append a new ball to the circle glyph as it traverses downward along with a downward zig-zagging line. Once the ball/line gets to the bottom, the coordinates reset and starts to drop from the top again. So, the position of the balls/circles should match the position of the lines and they are updated at the same time. For some unknown reason, the position of the balls/circles is getting out of sync with the lines and I can't figure out why. 

Note that things seem to behave as expected when only one of the glyphs (either circle OR line) is updated but not when both are updated simultaneously (i.e., if you comment out all of the lines with references to `ball`).
- **Self-contained example code that generates the issue (if possible)**

<pre>
from bokeh.models import Range1d
from bokeh.plotting import figure, show
from bokeh.client import push_session
from bokeh.io import curdoc
import random

session = push_session(curdoc())
n = 10
p = figure(x_range=Range1d(-n, n), y_range=Range1d(-1, n+1))
p.line([0], [n], color='red', line_width=3, name='line_1')
p.circle([0], [n], color='red', radius=0.3, name='ball_1')

def update(n=10):
    ball_1 = p.select_one({'name': 'ball_1'})
    line_1 = p.select_one({'name': 'line_1'})

    sign = random.choice([-1, 1])
    ball_1_x = list(ball_1.data_source.data['x'])
    ball_1_y = list(ball_1.data_source.data['y'])
    line_1_x = list(line_1.data_source.data['x'])
    line_1_y = list(line_1.data_source.data['y'])
         
    if ball_1_y != line_1_y:
        # I should never be in here
        print ball_1_y, line_1_y
        
    #if line_1_y[-1] > 1:
    if len(line_1_y) < 10:
        x = line_1_x[-1] + sign
        y = line_1_y[-1] - 1
        line_1_x.append(x)
        line_1_y.append(y)
        ball_1_x.append(x)
        ball_1_y.append(y)
    else:
        if ball_1_y != [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]:
            # I should never be in here
            print "Ball", ball_1_x, ball_1_y
        if line_1_y != [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]:
            # I should never be in here
            print "Line:", line_1_x, line_1_y
        line_1_x = [0]
        line_1_y = [n]
        ball_1_x = [0]
        ball_1_y = [n]
    
    ball_1.data_source.data = {'x': ball_1_x, 'y': ball_1_y}
    line_1.data_source.data = {'x': line_1_x, 'y': line_1_y}

curdoc().add_periodic_callback(update, 1)
session.show(p)
session.loop_until_closed()
</pre>

- **Printed Output from Code** 

<pre>
[10] [10, 9, 8, 7, 6]
[10, 5] [10, 9, 8, 7, 6, 5]
[10, 5, 4] [10, 9, 8, 7, 6, 5, 4]
[10, 9, 8, 7, 6] [10, 9, 8, 7, 6, 5, 4, 3]
[10, 9, 8, 7, 6, 2] [10, 9, 8, 7, 6, 5, 4, 3, 2]
[10, 9, 8, 7, 6, 2, 1] [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Ball [0, -1, -2, -1, -2, -2, -3] [10, 9, 8, 7, 6, 2, 1]
[10, 9] [10, 9, 8, 7]
[10, 9, 6] [10, 9, 8, 7, 6]
[10, 9, 6, 5] [10, 9, 8, 7, 6, 5]
[10, 9, 6, 5, 4] [10, 9, 8, 7, 6, 5, 4]
[10, 9, 6, 5, 4, 3] [10, 9, 8, 7, 6, 5, 4, 3]
[10, 9, 6, 5, 4, 3, 2] [10, 9, 8, 7, 6, 5, 4, 3, 2]
[10, 9, 6, 5, 4, 3, 2, 1] [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Ball [0, 1, 2, 1, 2, 1, 0, 1] [10, 9, 6, 5, 4, 3, 2, 1]
[10, 9, 8, 7] [10]
[10, 9, 8, 7, 9] [10, 9]
[10, 9, 8, 7, 9, 8] [10, 9, 8]
[10, 9, 8, 7, 9, 8, 7] [10, 9, 8, 7]
[10, 9, 8, 7, 9, 8, 7, 6] [10, 9, 8, 7, 6]
[10, 9, 8, 7, 9, 8, 7, 6, 5] [10, 9, 8, 7, 6, 5]
[10, 9, 8, 7, 9, 8, 7, 6, 5, 4] [10, 9, 8, 7, 6, 5, 4]
[10, 9, 8, 7, 9, 8, 7, 6, 5, 4, 3] [10, 9, 8, 7, 6, 5, 4, 3]
[10, 9, 8, 7, 9, 8, 7, 6, 5, 4, 3, 2] [10, 9, 8, 7, 6, 5, 4, 3, 2]
[10, 9, 8, 7, 9, 8, 7, 6, 5, 4, 3, 2, 1] [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
Ball [0, 1, 2, 1, -1, 0, 1, 2, 1, 2, 3, 2, 1] [10, 9, 8, 7, 9, 8, 7, 6, 5, 4, 3, 2, 1]
</pre>

- **Software version information output by `bokeh info` command**

<pre>
Python version      :  2.7.11 |Anaconda 2.5.0 (x86_64)| (default, Jun 15 2016, 16:09:16) 
IPython version     :  4.0.3
Bokeh version       :  0.12.0
BokehJS static path :  /path/to/anaconda/lib/python2.7/site-packages/bokeh/server/static
</pre>

**Comment by seanlaw** [comment]
@bryevdv This is part of the Galton Board example I was talking about working on...

**Comment by bryevdv** [comment]
`1 ms` is very fast update interval. It's actually after than the redraw throttling. Does this also happen if you set a slower update interval, say `100 ms`?

**Comment by seanlaw** [comment]
Yes, it looks like 100ms is okay and does not have problems updating the glyphs. It isn't entirely clear to me why the update interval is a problem. The lines update correctly but the balls don't. In fact, the balls are found to update as [10, 9, 6, 5] (instead of [10, 9, 8, 7]). Not sure what happened to the "8, 7".

**Comment by bryevdv** [comment]
Every property change triggers a new update message. There is no notion of batching these events, at least not yet. My guess is that such a short update interval is exposing some kind of race condition where the server's response to an update is coming in during the next python update. What are your actual requirements? As I mentioned right now there is a BokehJS throttling rate of around 20-30ms, so there is really no point int ever having an update rate faster than that. It's possible the throttle rate could be made configurable, but at some point you run into how fast the browser can draw and it's never going to make sense to send data faster than that.

**Comment by bryevdv** [comment]
I should add, the existing protocol would have no problem batching events, there's just no way to suppress the immediate protocol message sending. But it might not be too bad. @havocp what do you think about a decorator for callbacks? If it's present then all the changes are collected and then only when the function returns does one big message get generated and sent?

**Comment by havocp** [comment]
Isn't the callback running too often, so we need to batch updates across multiple invocations of the update() callback?

I think the general problem here may not be batching but backlog... that is, one side of the connection going faster than the other one can keep up, so the queue of events grows without bound.

One outline of a solution is "backpressure" - client says "I am ready to handle events, send me up to 5" and the server does that and then _pauses_ until the client sends a new message "I am ready for 1 more" - the server probably has to stop running callbacks when it's maxed out what the client has requested, in bokeh's case.  The same has to happen in the other direction (client pauses if server is busy). The idea is to make the queues finite on both sides by not doing work that could add to them until the peer drains some stuff so it can handle it. At least, I _think_ that's the general idea, I have not dug into it too deeply; backpressure is a topic with some body of literature available I think.

An even simpler approach would be to only send the next event when getting an ack for the prior event, but the problem is that would be too slow; some kind of buffering is needed, it just needs to be a finite buffer.

We do know about all callbacks since we don't expose tornado directly (at least not easily) so it should be possible to pause a session when a client gets lagged.

App-global (outside-of-session) callbacks from server_lifecycle.py would have to manually check for paused sessions or something.

For batching it might be simple to keep a 1-event buffer for a few tens of millis and collapse adjacent changes to the exact same property on the exact same model... but I think stuff could break if we collapse changes that aren't adjacent? In most real world cases that seems likely, that we'd change several different properties in each callback. Batching could break something then maybe... maybe not. Reordering events seems like a thing to worry about though.

**Comment by bryevdv** [comment]
This is implemented in #6941 with `doc.hold()` and `doc.unhold()` which can be called to suppress and then re-enable model change events and callbacks for a duration. Tho FWIW the original problem seems to have been fixed some time before, with other work, since the code (lightly updated) works both with and without the PR. Here is the updated script:
```
from bokeh.models import Range1d
from bokeh.plotting import figure, show
from bokeh.io import curdoc
import random

n = 10

p = figure(x_range=Range1d(-n, n), y_range=Range1d(-1, n+1))

ball = p.line([0], [n], color='red', line_width=3)
line = p.circle([0], [n], color='red', radius=0.3)

def update(n=10):
    sign = random.choice([-1, 1])
    ball_x = ball.data_source.data['x']
    ball_y = ball.data_source.data['y']
    line_x = line.data_source.data['x']
    line_y = line.data_source.data['y']

    #if line_y[-1] > 1:
    if len(line_y) < 10:
        x = line_x[-1] + sign
        y = line_y[-1] - 1
        line_x.append(x)
        line_y.append(y)
        ball_x.append(x)
        ball_y.append(y)
    else:
        line_x = [0]
        line_y = [n]
        ball_x = [0]
        ball_y = [n]

    curdoc().hold()
    ball.data_source.data = {'x': ball_x, 'y': ball_y}
    line.data_source.data = {'x': line_x, 'y': line_y}
    curdoc().unhold()

curdoc().add_periodic_callback(update, 100)
curdoc().add_root(p)
```

**Comment by bryevdv** [comment]
![screenflow](https://user-images.githubusercontent.com/1078448/30524230-a0d727a0-9bbd-11e7-8ff7-5d79bd95fe97.gif)
